iree/hal/cuda/cuda_headers.h

```diff
@@ -9,4 +9,18 @@
 
 #include "cuda.h"  // IWYU pragma: export
 
+#if IREE_ENABLE_CUPTI  // CUPTI is only used for tracing.
```

Comment:
Is this define getting set? I think you'd need to set the `DEFINES` on `iree::hal::cuda::dynamic_symbols` (based on the `CUDA::cupti` target existing), rather than the `cuda_headers` target.

You could create a list with conditional entries to make that a bit cleaner:

```cmake
if(TARGET CUDA::cupti)
  set(_IREE_CUDA_HEADERS_DEFINES "IREE_ENABLE_CUPTI")
else()
  set(_IREE_CUDA_HEADERS_DEFINES "")
endif()

...

iree_cc_library(
  NAME
    dynamic_symbols

  DEFINES
    _IREE_CUDA_HEADERS_DEFINES 
```

Reply:
Oh, and that check should also include `IREE_ENABLE_RUNTIME_TRACING`. A downstream project may set up CUDA targets but not want IREE's runtime tracing enabled.

Reply:
I'm not sure I follow. I've tested it and this is getting set when CUDA backend with tracing is enabled. Its getting set in `build_tools/third_party/cuda/CMakeList.txt` which is what builds this file:

```
if (TARGET CUDA::cupti)
  external_cc_library(
    PACKAGE
      cuda_headers
    NAME
      cuda_headers
    ROOT
      ${CUDA_HEADERS_API_ROOT}
    DEFINES
      IREE_ENABLE_CUPTI
    INCLUDES
      ${CUDA_HEADERS_API_ROOT}
      ${CUDA_INCLUDE_DIR}
      ${CUPTI_INCLUDE_DIR}
  )
```

Also I just realized I removed the definitions for `CUDA_INCLUDE_DIR` and `CUPTI_INCLUDE_DIR`. Added those back. With these put back, I think the way I have it is cleanest because it bundles the find path and the build rule definitions together. The two paths need separate variables because `find_path` doesn't support multiple paths in a single variable.

Reply:
Keeping the define close to where it is used seems like it would be easier to understand in the future.

`build_tools/third_party/` then just provides targets, header files, etc. and the code using that elsewhere in the project can decide what it wants to do with it.

Reply:
Per our discussion on team chat: The build files for third_party can't go in third_party because they're not third_party code, so we have to have this unusual split between custom build files and the third_party source.